video/vdu_buffered.h

```diff
@@ -1060,4 +1072,203 @@ void VDUStreamProcessor::bufferCopyAndConsolidate(uint16_t bufferId, std::vector
+				cd.input_count, cd.output_count, pct, destination);
+	} else {
+		debug_log("bufferCompress: cannot allocate temporary buffer of %d bytes\n\r", COMPRESSION_OUTPUT_CHUNK_SIZE);
+	}
+}
+
+// VDU 23, 0, &A0, bufferId; &1D, sourceBufferId; sourceBufferId; ...; 65535; : Decompress blocks from buffers
+// Decompress (blocks from) a list of buffers into a new buffer.
+// List is terminated with a bufferId of 65535 (-1).
+// Replaces the target buffer with the new one.
+// If target buffer is included in the source list it will be skipped.
+//
+void VDUStreamProcessor::bufferDecompress(uint16_t bufferId, uint16_t sourceBufferId) {
+	// Validate the compression header
+	if (buffers[sourceBufferId][0]->size() < sizeof(CompressionFileHeader)) {
+		debug_log("bufferDecompress: buffer too small for header\n\r");
+		return;
+	}
+	
+	auto p_hdr = (const CompressionFileHeader*) buffers[sourceBufferId][0]->getBuffer();
+	if (p_hdr->marker[0] != 'C' ||
+		p_hdr->marker[1] != 'm' ||
+		p_hdr->marker[2] != 'p' ||
+		p_hdr->type != COMPRESSION_TYPE_TURBO) {
+		debug_log("bufferDecompress: header is invalid\n\r");
+		return;
+	}
+	auto orig_size = p_hdr->orig_size;
+
+	buffers[bufferId].clear();
+	debug_log("Decompressing into buffer %u\n\r", bufferId);
+
+	// create a temporary output buffer, which may be expanded during decompression
+	uint8_t* p_temp = (uint8_t*) ps_malloc(COMPRESSION_OUTPUT_CHUNK_SIZE);
+	if (p_temp) {
+		// prepare for doing compression
+		DecompressionData dd;
+		agon_init_decompression(&dd, &p_temp, &local_write_decompressed_byte);
+
+		// loop thru blocks stored against the source buffer ID
+		uint32_t skip_hdr = sizeof(CompressionFileHeader);
+		dd.input_count = skip_hdr;
+		for (const auto &block : buffers[sourceBufferId]) {
+			// decompress the block into our temporary buffer
+			auto bufferLength = block->size() - skip_hdr;
+			auto p_data = block->getBuffer();
+			debug_log(" from buffer %u [%08X] %u bytes\n\r", sourceBufferId, p_data, bufferLength);
+			debug_log(" %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX\n\r",
+						p_data[0], p_data[1], p_data[2], p_data[3],
+						p_data[4], p_data[5], p_data[6], p_data[7],
+						p_data[8], p_data[9], p_data[10], p_data[11]);
+			p_data += skip_hdr;
+			skip_hdr = 0;
+			dd.input_count += bufferLength;
+			while (bufferLength--) {
+				agon_decompress_byte(&dd, *p_data++);
+			}
+		}
+
+		// make a single buffer with all of the temporary output data
+		auto bufferStream = make_shared_psram<BufferStream>(dd.output_count);
+		if (!bufferStream || !bufferStream->getBuffer()) {
+			// buffer couldn't be created
+			debug_log("bufferDecompress: failed to create buffer %d\n\r", bufferId);
+			return;
+		}
+		auto destination = bufferStream->getBuffer();
+		debug_log(" %02hX %02hX %02hX %02hX\n\r",
+					p_temp[0], p_temp[1], p_temp[2], p_temp[3]);
+		memcpy(destination, p_temp, dd.output_count);
+		heap_caps_free(p_temp);
+		buffers[bufferId].push_back(bufferStream);
```

Comment:
we could make this more elegant (and simpler) if we just provide a way of initialising a buffer stream with a pre-allocated chunk of memory, i.e. pass in `dd->context`

if the output block were already allocated in psram then this would also remove the need to do a memcpy

Reply:
This was kind of a leftover from before we had the header, telling what the original uncompressed size was. Now that we have it, sure, we could pre-allocate.